### DISCUSSION TITLE
A few Questions

### DISCUSSION THREAD

**Discussion by crypto-dot** [RayTracing/raytracing.github.io#920] [Q&A] [answered]
I am curious on what the reason is for multiplying the variable u by the horizonal vector and adding the variable v times the vertical vector. Is this equivalent to moving the offset vectors by some proportion of u and v in relation to the image size? Also why does gradient change in the y and x direction, the book mentioned that normalizing the vector has something to do with this but I can't seem to wrap my head around why. Additionally why subtract the origin from the lower_left_corner (Esp. if the origin is 0,0,0)?
   
```
    for (int j = image_height-1; j >= 0; --j) {
        std::cerr << "\rScanlines remaining: " << j << ' ' << std::flush;
        for (int i = 0; i < image_width; ++i) {
            auto u = double(i) / (image_width-1);
            auto v = double(j) / (image_height-1);
            ray r(origin, lower_left_corner + u*horizontal + v*vertical - origin);
            color pixel_color = ray_color(r);
            write_color(std::cout, pixel_color);
        }
    }
```

**Comment by hollasch** [comment, marked as answer]
The "origin" is the origin of the scene rays. This ray origin can be anywhere in the scene; its 3D coordinates are arbitrary, and need not be (0,0,0).

After that, we want vectors that go from this camera origin to some 3D pixel coordinate on the virtual image viewport, so we need a direction vector that does that. B - A yields the vector from point A to point B, and pixel coordinate - ray origin gives you the vector from the camera origin (focal point) to the image pixel. That's why we subtract the ray origin point.

`horizontal` is the vector from the image viewport lower left corner to the lower _right_ corner. `vertical` is the vector from the image lower left corner to the _upper_ left corner. Given that, we just want a value from 0 to 1 that scales these two vectors as we traverse the image pixels. (u,v) of (0.5, 0.5) gives us the image point that halfway to the right side, halfway up.

Thus, `lower_left_corner + u*horizontal + v*vertical` gets us the coordinates of the image pixel on our virtual viewport (a rectangle in space), and subtracting the ray origin gives us the direction from the camera focal point to the image pixel.

**Reply by crypto-dot** [reply]
Thank you. It seems I was confusing what was actually a point vs what was actually a vector and the difference between the two.
